modin/engines/modin_dataframe.py

```diff
@@ -0,0 +1,477 @@
+# copyright ownership.  The Modin Development Team licenses this file to you under the
+# Apache License, Version 2.0 (the "License"); you may not use this file except in
+# compliance with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+from abc import ABC, abstractmethod
+from typing import List, Hashable, Optional, Callable, Union, Dict
+
+
+class ModinDataframe(ABC):
+    """An abstract class that represents the Parent class for any DataFrame class.
+
+    This class is intended to specify the behaviours that a DataFrame must implement.
+
+    For more details about how these methods were chosen, please refer to this
+    (http://www.vldb.org/pvldb/vol13/p2033-petersohn.pdf) paper, which specifies
+    a DataFrame algebra that this class exposes.
+
+    """
+
+    @abstractmethod
+    def mask(
+        self,
+        row_labels: Optional[List[Hashable]] = None,
+        row_positions: Optional[List[int]] = None,
+        col_labels: Optional[List[Hashable]] = None,
+        col_positions: Optional[List[int]] = None,
+    ) -> "ModinDataframe":
+        """Mask rows and columns in the dataframe.
+
+        Allow users to perform selection and projection on the row and column number (positional notation),
+        in addition to the row and column labels (named notation)
+
+        Parameters
+        ----------
+        row_labels : list of hashable
+            The row labels to extract.
+        row_positions : list of int
+            The row indices to extract.
+        col_labels : list of hashable
+            The column labels to extract.
+        col_positions : list of int
+            The column indices to extract.
+
+        Returns
+        -------
+        ModinDataframe
+             A new ModinDataframe from the mask provided.
+        """
+        pass
+
+    @abstractmethod
+    def filter_by_types(self, types: List[Hashable]) -> "ModinDataframe":
+        """Allow the user to specify a type or set of types by which to filter the columns.
+
+        Parameters
+        ----------
+        types: list of hashables
+            The types to filter columns by.
+
+        Returns
+        -------
+        ModinDataframe
+             A new ModinDataframe from the filter provided.
+        """
+        pass
+
+    @abstractmethod
+    def map(
+        self,
+        function: Callable,
+        axis: Optional[int] = None,
+        result_schema: Optional[Dict[Hashable, type]] = None,
+    ) -> "ModinDataframe":
+        """Apply a user-defined function row- wise (or column-wise if axis=1).
+
+        Notes
```

Comment:
I would prefer `Notes` follows after `Returns` as it is done in other projects. Also, numpydoc guide describes `Notes` as one of the last sections in a docstring.

Reply:
+1 let's follow numpydoc standard